### PR TITLE
LocalAsyncCache when loader throws and exception do not cache results

### PR DESCRIPTION
**Pull request by lifey**
(Removed main from tests)

### CODE REVIEW

File: ob1k-cache/src/main/java/com/outbrain/ob1k/cache/LocalAsyncCache.java [hunk 1]
```diff
@@ -112,9 +112,22 @@ public LocalAsyncCache() {
 
   private ComposableFuture<V> loadElement(final K key) {
     final ComposableFuture<V> loadedElement = loader.load(cacheName, key).materialize();
-    return loadedElement.continueOnError((FutureErrorHandler<V>) error -> {
-      loadingCache.asMap().remove(key, loadedElement);
-      return fromError(error);
+    return loadedElement;
+  }
+
+  private ComposableFuture<V> extractMapEntry(ComposableFuture<Map<K, V>> loaded ,K key) {
```
Comment (marenzo) [inline ob1k-cache/src/main/java/com/outbrain/ob1k/cache/LocalAsyncCache.java]: finals & code styling

Comment (marenzo) [inline ob1k-cache/src/main/java/com/outbrain/ob1k/cache/LocalAsyncCache.java]: also, maybe call it "extractCacheResultEntry"?

File: ob1k-cache/src/main/java/com/outbrain/ob1k/cache/LocalAsyncCache.java [hunk 2]
```diff
@@ -112,9 +112,22 @@ public LocalAsyncCache() {
...
+    return loaded
+      .continueOnSuccess((FutureSuccessHandler<Map<K, V>, V>) loaderResults -> {
+      final V res = loaderResults.get(key);
+      if (res != null) {
+        return fromValue(res);
+      } else {
+        if (failOnMissingEntries) {
```
Comment (marenzo) [inline ob1k-cache/src/main/java/com/outbrain/ob1k/cache/LocalAsyncCache.java]: failOnMissingEntries is always true, and cannot be changed outside. Maybe just removing this redundant flag?

File: ob1k-cache/src/main/java/com/outbrain/ob1k/cache/LocalAsyncCache.java (hunk 1 above)
Comment (marenzo) [inline ob1k-cache/src/main/java/com/outbrain/ob1k/cache/LocalAsyncCache.java]: and last, but this is just mine opinion: you shouldn't accept the future as a param, rather just return the FutureSuccessHandler, and do the continuation outside.

File: ob1k-cache/src/main/java/com/outbrain/ob1k/cache/LocalAsyncCache.java [hunk 3]
```diff
@@ -147,11 +144,14 @@ public LocalAsyncCache() {
   public ComposableFuture<V> getAsync(final K key) {
     try {
       if (loadingCache != null) {
-        ComposableFuture<V> res = loadingCache.get(key);
+        final ComposableFuture<V> res = loadingCache.get(key);
         if (res == null) {
-          res = fromValue(null);
+          return fromValue(null);
```
Comment (marenzo) [inline ob1k-cache/src/main/java/com/outbrain/ob1k/cache/LocalAsyncCache.java]: fromNull

File: ob1k-cache/src/main/java/com/outbrain/ob1k/cache/LocalAsyncCache.java [hunk 4]
```diff
@@ -164,12 +164,23 @@ public LocalAsyncCache() {
...
+          loadingCache.asMap().remove(key);
+        }
+      });
+    }
+  }
+
   @Override
   public ComposableFuture<Map<K, V>> getBulkAsync(final Iterable<? extends K> keys) {
     try {
       final ImmutableMap<K, ComposableFuture<V>> innerMap;
```
Comment (marenzo) [inline ob1k-cache/src/main/java/com/outbrain/ob1k/cache/LocalAsyncCache.java]: can it be uppercasted to Map?

File: ob1k-cache/src/main/java/com/outbrain/ob1k/cache/LocalAsyncCache.java [hunk 5]
```diff
@@ -164,12 +164,23 @@ public LocalAsyncCache() {
     }
   }
 
+  void unloadErrorsFromCache(final ImmutableMap<K, ComposableFuture<V>> innerMap, final Iterable<? extends K> keys) {
```
Comment (marenzo) [inline ob1k-cache/src/main/java/com/outbrain/ob1k/cache/LocalAsyncCache.java]: some modifiers? :)

File: ob1k-cache/src/main/java/com/outbrain/ob1k/cache/LocalAsyncCache.java (hunk 2 above)
Comment (lifey) [inline ob1k-cache/src/main/java/com/outbrain/ob1k/cache/LocalAsyncCache.java]: there is a constructor for that

File: ob1k-cache/src/main/java/com/outbrain/ob1k/cache/LocalAsyncCache.java (hunk 4 above)
Comment (lifey) [inline ob1k-cache/src/main/java/com/outbrain/ob1k/cache/LocalAsyncCache.java]: not needed in this scope

**Comment by marenzo** [comment]
looks good

**Pull request merged**